rusk/src/lib/chain/rusk.rs

```diff
@@ -494,16 +565,43 @@ fn accept(
 fn execute(
     session: &mut Session,
     tx: &PhoenixTransaction,
+    gas_per_deploy_byte: Option<u64>,
 ) -> Result<CallReceipt<Result<Vec<u8>, ContractError>>, PiecrustError> {
+    // Transaction will be discarded if it is a deployment transaction
+    // with gas limit smaller than deploy charge.
+    if let Some(deploy) = tx.payload().contract_deploy() {
+        let deploy_charge =
+            bytecode_charge(&deploy.bytecode, &gas_per_deploy_byte);
+        if tx.payload().fee.gas_limit < deploy_charge {
+            return Err(PiecrustError::Panic(
+                "not enough gas to deploy".into(),
+            ));
+        }
+    }
+
+    let tx_stripped = tx.strip_off_bytecode();
     // Spend the inputs and execute the call. If this errors the transaction is
     // unspendable.
     let mut receipt = session.call::<_, Result<Vec<u8>, ContractError>>(
         TRANSFER_CONTRACT,
         "spend_and_execute",
-        tx,
+        tx_stripped.as_ref().unwrap_or(tx),
         tx.payload().fee.gas_limit,
     )?;
 
+    // Deploy if this is a deployment transaction and spend part is successful.
+    if let Some(deploy) = tx.payload().contract_deploy() {
+        if receipt.data.is_ok() {
```

Comment:
Just for my understanding... this `if` is here because an error in the contract call consumes all gas after? (now line 606)

Reply:
yes, exactly, if `receipt.data` is not OK, we should not execute deployment, as the only thing we should do in this situation is to consume all gas limit, call refund (not refunding anything) and return